polyfill/lib/calendar.mjs

```diff
@@ -333,53 +332,1354 @@ function resolveNonLunisolarMonth(calendarDate) {
+        // names that contain non-ASCII characters like Japanese eras. Also
+        // ignore extra content in parentheses like JPN era date ranges.
+        value = value.split(' (')[0];
+        result.era = value
+          .normalize('NFD')
+          .replace(/[^-0-9 \p{L}]/gu, '')
+          .replace(' ', '-')
+          .toLowerCase();
+      }
+    }
+    if (result.eraYear === undefined) {
+      // Node 12 has outdated ICU data that lacks the `relatedYear` field in the
+      // output of Intl.DateTimeFormat.formatToParts.
+      throw new RangeError(
+        `Intl.DateTimeFormat.formatToParts lacks relatedYear in ${this.id} calendar. Try Node 14+ or modern browsers.`
       );
     }
-    if (fields.eraYear === undefined) {
-      return;
+    // Handle pre-Meiji Japanese eras that are present in Intl output but not
+    // yet (maybe never) used by Temporal (see #526)
+    if (this.reviseIntlEra) {
+      const { era, eraYear } = this.reviseIntlEra(result, isoDate);
+      result.era = era;
+      result.eraYear = eraYear;
+    }
+    if (this.checkJulianBug) this.checkJulianBug(result, isoDate);
+
+    const calendarDate = this.adjustCalendarDate(result, cache, 'constrain', true);
+    if (calendarDate.year === undefined) throw new RangeError(`Missing year converting ${JSON.stringify(isoDate)}`);
+    if (calendarDate.month === undefined) throw new RangeError(`Missing month converting ${JSON.stringify(isoDate)}`);
+    if (calendarDate.day === undefined) throw new RangeError(`Missing day converting ${JSON.stringify(isoDate)}`);
+    cache.set(key, calendarDate);
+    // Also cache the reverse mapping
+    ['constrain', 'reject'].forEach((overflow) => {
+      const keyReverse = JSON.stringify({
+        func: 'calendarToIsoDate',
+        year: calendarDate.year,
+        month: calendarDate.month,
+        day: calendarDate.day,
+        overflow,
+        id: this.id
+      });
+      cache.set(keyReverse, isoDate);
+    });
+    return calendarDate;
+  },
+  validateCalendarDate(calendarDate) {
+    let { month, year, day, era, eraYear, monthCode, monthExtra } = calendarDate;
+    // When there's a suffix (e.g. "5bis" for a leap month in Chinese calendar)
+    // the derived class must deal with it.
+    if (monthExtra !== undefined) throw new RangeError('Unexpected `monthExtra` value');
+    if (year === undefined && eraYear === undefined) throw new TypeError('year or eraYear is required');
+    if (month === undefined && monthCode === undefined) throw new TypeError('month or monthCode is required');
+    if (day === undefined) throw new RangeError('Missing day');
+    if (eraYear !== undefined && era === undefined && this.hasEra) throw new TypeError('era is required');
+    if (era !== undefined && !this.hasEra) throw new RangeError(`No eras in ${this.id} calendar`);
+    if (era !== undefined && eraYear === undefined && this.hasEra) throw new TypeError('eraYear is required');
```

Comment:
I agree with all of this suggestion except this line: 
```js
if (era !== undefined || eraYear !== undefined && !this.hasEra) throw new RangeError(`No eras in ${this.id} calendar`);
```

If the calendar doesn't use eras, then it seems safe to just ignore `era` and `eraYear` completely. Not sure we need to throw an exception for a unit that's ignored by the calendar. What do you think?

Reply:
Oh, I just realized that the era test was my own previous code, just relocated and refactored. ;-)  Regardless, it still seems OK to ignore eras here because that's what ISO does, e.g. 
`Temporal.PlainDate.from({era: 'bc', year: 1, month: 1, day: 1})` doesn't throw. 

I'd also be OK with throwing here, as long as ISO does too... but that's a spec change to 262 and I'm not sure this meets the bar. What do you think?

Regardless, my strong opinion is that either all built-in calendars (including ISO) should do this check, or none should.

Reply:
Ah, agree it doesn't meet the bar for a spec change and that it should be consistent. Let's ignore, then. But maybe the era & eraYear consistency checks (in part because they're also already enforced by PrepareTemporalFields) should only happen here if `this.hasEra`, so that a calendar without eras isn't bothering to check them at all? That sounds consistent.

Reply:
Yep, agreed. I left those checks in including your changes to the messages. Take a look at this code in my latest checkin, and feel free to re-open this conversation if you disagree with how it ended up.

Reply:
Oh, right, they already do that.

/me looks forlornly at empty coffee cup